File: audiocontrol/audio.py
import subprocess
import re

class AudioManager:
    @staticmethod
    def get_audio_devices():
        devices = []
        try:
            result = subprocess.run(['wpctl', 'status'], capture_output=True, text=True)
            output = result.stdout
        except Exception as e:
            print(f"Error running wpctl: {e}")
            return []

        lines = output.splitlines()
        in_sinks = False
        
        # Regex to match sink lines:
        # Optional indent, optional *, ID, dot, Name, [vol: ...]
        # Example: │  *   64. Navi 10 HDMI Audio Digital Stereo (HDMI) [vol: 0.40]
        sink_pattern = re.compile(r'^\s*│\s*(\*)?\s*(\d+)\.\s+(.+?)\s+\[vol:\s*([\d\.]+)')

        for line in lines:
            if "Sinks:" in line:
                in_sinks = True
                continue
            if in_sinks and not line.strip().startswith("│"):
                # End of sinks section (empty line or next section)
                if line.strip() == "":
                    continue # could be empty line inside? usually structure is strict
                if "Sources:" in line or "Filters:" in line or "Streams:" in line:
                    in_sinks = False
                    break
            
            if in_sinks:
                match = sink_pattern.search(line)
                if match:
                    is_active = bool(match.group(1))
                    device_id = match.group(2)
                    name = match.group(3).strip()
                    volume_str = match.group(4)
                    volume = float(volume_str) if volume_str else 0.0
                    
                    # Icon heuristic
                    icon_name = "audio-speakers-symbolic"
                    lower_name = name.lower()
                    if "headphone" in lower_name:
                        icon_name = "audio-headphones-symbolic"
                    elif "hdmi" in lower_name or "tv" in lower_name or "display" in lower_name:
                        icon_name = "video-display-symbolic"
                    elif "bluetooth" in lower_name:
                        icon_name = "bluetooth-active-symbolic"
                    
                    devices.append((name, icon_name, is_active, volume, device_id))
        
        return devices

    @staticmethod
    def set_volume(device_id, value):
        try:
            # Set absolute volume
            subprocess.run(['wpctl', 'set-volume', str(device_id), str(value)], check=False)
        except Exception as e:
            print(f"Failed to set volume: {e}")

    @staticmethod
    def set_default_device(device_id):
        try:
            subprocess.run(['wpctl', 'set-default', str(device_id)], check=False)
        except Exception as e:
            print(f"Failed to set default device: {e}")

File: src/audio.c
#define _POSIX_C_SOURCE 200809L
#include "audio.h"
#include <ctype.h>
#include <regex.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_DEVICES 32
#define LINE_BUF_SIZE 1024

static void get_icon_name(const char *name, char *icon_name, size_t max_len) {
  char lower_name[256];
  strncpy(lower_name, name, sizeof(lower_name) - 1);
  lower_name[sizeof(lower_name) - 1] = '\0';

  // Convert to lowercase
  for (int i = 0; lower_name[i]; i++) {
    lower_name[i] = tolower(lower_name[i]);
  }

  if (strstr(lower_name, "headphone")) {
    strncpy(icon_name, "audio-headphones-symbolic", max_len);
  } else if (strstr(lower_name, "hdmi") || strstr(lower_name, "tv") ||
             strstr(lower_name, "display")) {
    strncpy(icon_name, "video-display-symbolic", max_len);
  } else if (strstr(lower_name, "bluetooth")) {
    strncpy(icon_name, "bluetooth-active-symbolic", max_len);
  } else {
    strncpy(icon_name, "audio-speakers-symbolic", max_len);
  }
}

AudioDevice *get_audio_devices(int *count) {
  FILE *fp;
  char line[LINE_BUF_SIZE];
  AudioDevice *devices = malloc(sizeof(AudioDevice) * (MAX_DEVICES + 1));
  *count = 0;

  if (!devices)
    return NULL;

  fp = popen("wpctl status", "r");
  if (fp == NULL) {
    free(devices);
    return NULL;
  }

  bool in_sinks = false;
  regex_t regex;
  // Regex: ^\s*│\s*(\*)?\s*(\d+)\.\s+(.+?)\s+\[vol:\s*([\d\.]+)
  // C string escaping makes this ugly
  const char *pattern =
      "^[[:space:]]*│[[:space:]]*(\\*?)[[:space:]]*([0-9]+)\\.[[:space:]]+(.+?)"
      "[[:space:]]+\\[vol:[[:space:]]*([0-9\\.]+)";

  if (regcomp(&regex, pattern, REG_EXTENDED) != 0) {
    pclose(fp);
    free(devices);
    return NULL;
  }

  while (fgets(line, sizeof(line), fp) != NULL) {
    if (strstr(line, "Sinks:")) {
      in_sinks = true;
      continue;
    }

    if (in_sinks) {
      // Check for end of section
      // If line doesn't start with │ (ignoring whitespace), it's likely end of
      // section But we need to be careful. The regex expects │. Let's check if
      // it contains "Sources:" or "Filters:" or "Streams:"
      if (strstr(line, "Sources:") || strstr(line, "Filters:") ||
          strstr(line, "Streams:")) {
        in_sinks = false;
        break;
      }

      // Simple check for empty line or line without pipe might be enough,
      // but let's rely on regex matching for valid lines.

      regmatch_t matches[5];
      if (regexec(&regex, line, 5, matches, 0) == 0) {
        if (*count >= MAX_DEVICES)
          break;

        AudioDevice *dev = &devices[*count];

        // Group 1: "*" or empty (Active)
        dev->is_active =
            (matches[1].rm_so != -1 && matches[1].rm_eo > matches[1].rm_so);

        // Group 2: ID
        char id_str[16];
        int len = matches[2].rm_eo - matches[2].rm_so;
        if (len >= (int)sizeof(id_str))
          len = sizeof(id_str) - 1;
        strncpy(id_str, line + matches[2].rm_so, len);
        id_str[len] = '\0';
        dev->id = atoi(id_str);

        // Group 3: Name
        len = matches[3].rm_eo - matches[3].rm_so;
        if (len >= (int)sizeof(dev->name))
          len = sizeof(dev->name) - 1;
        strncpy(dev->name, line + matches[3].rm_so, len);
        dev->name[len] = '\0';

        // Group 4: Volume
        char vol_str[16];
        len = matches[4].rm_eo - matches[4].rm_so;
        if (len >= (int)sizeof(vol_str))
          len = sizeof(vol_str) - 1;
        strncpy(vol_str, line + matches[4].rm_so, len);
        vol_str[len] = '\0';
        dev->volume = atof(vol_str);

        get_icon_name(dev->name, dev->icon_name, sizeof(dev->icon_name));

        (*count)++;
      }
    }
  }

  regfree(&regex);
  pclose(fp);

  // Terminator
  devices[*count].id = -1;
  return devices;
}

void set_volume(int device_id, double value) {
  char cmd[64];
  snprintf(cmd, sizeof(cmd), "wpctl set-volume %d %.2f", device_id, value);
  system(cmd);
}

void set_default_device(int device_id) {
  char cmd[64];
  snprintf(cmd, sizeof(cmd), "wpctl set-default %d", device_id);
  system(cmd);
}
